src/Orleans/Runtime/CallbackData.cs

```diff
@@ -153,6 +153,43 @@ public void OnTimeout()
             callback(error, context);
         }
 
+        public void Fail()
```

Comment:
probably a good idea to write a common function for OnTimeout and Fail and just pass the exception inside and what ever difference exists between the two, do based on the exception type. Otherwise, too much code duplication.